File: 270-Closest-Binary-Search-Tree-Value/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int closestValue(TreeNode root, double target) {
        double closest = Math.abs(root.val - target);
        int value = root.val;
        TreeNode current = root;
        while (current != null) {
            if (compare(closest, Math.abs(current.val-target)) > 0) {
                closest = Math.abs(current.val-target);
                value = current.val;
            }
            if (compare(current.val, target) < 0)
                current = current.right;
            else if(compare(current.val, target) > 0)
                current = current.left;
            else break;
        }
        return value; 
    }
    
    int compare(double val1, double val2) {
        double temp = val1 - val2;
        if(temp < -1e-10) return -1;
        if(temp > 1e-10) return 1;
        return 0;
    }
}

File: 270-Closest-Binary-Search-Tree-Value/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        if not root:
            return -1
        result = root.val
        diff = abs(root.val - target)
        cur = root
        while cur:
            if cur.val == target:
                return cur.val
            if abs(cur.val - target) < diff:
                diff = abs(target - cur.val)
                result = cur.val
            if cur.val < target:
                cur = cur.right
            else:
                cur = cur.left
        return result